src/workerd/api/node/streams-test.js

```diff
@@ -11234,3 +11234,799 @@ export const push_strings = {
+          cb();
+          writeCalled.resolve();
+        }
+      });
+      passThrough.pipe(dest);
+      passThrough.pipe(dest);
+      strictEqual(passThrough._events.data.length, 2);
+      strictEqual(passThrough._readableState.pipes.length, 2);
+      strictEqual(passThrough._readableState.pipes[0], dest);
+      strictEqual(passThrough._readableState.pipes[1], dest);
+      passThrough.unpipe(dest);
+      strictEqual(passThrough._events.data.length, 1);
+      strictEqual(passThrough._readableState.pipes.length, 1);
+      deepStrictEqual(passThrough._readableState.pipes, [dest]);
+      passThrough.write('foobar');
+      passThrough.pipe(dest);
+      await writeCalled.promise;
+    }
+    {
+      const passThrough = new PassThrough();
+      const writeCalled = deferredPromise();
+      let writeCount = 0;
+      const dest = new Writable({
+        write: (chunk, encoding, cb) => {
+          strictEqual(`${chunk}`, 'foobar');
+          cb();
+          if (++writeCount == 2) writeCalled.resolve();
+        }
+      });
+      passThrough.pipe(dest);
+      passThrough.pipe(dest);
+      strictEqual(passThrough._events.data.length, 2);
+      strictEqual(passThrough._readableState.pipes.length, 2);
+      strictEqual(passThrough._readableState.pipes[0], dest);
+      strictEqual(passThrough._readableState.pipes[1], dest);
+      passThrough.write('foobar');
+      await writeCalled.promise;
+    }
+    {
+      const passThrough = new PassThrough();
+      const dest = new Writable({
+        write: fail
+      });
+      passThrough.pipe(dest);
+      passThrough.pipe(dest);
+      strictEqual(passThrough._events.data.length, 2);
+      strictEqual(passThrough._readableState.pipes.length, 2);
+      strictEqual(passThrough._readableState.pipes[0], dest);
+      strictEqual(passThrough._readableState.pipes[1], dest);
+      passThrough.unpipe(dest);
+      passThrough.unpipe(dest);
+      strictEqual(passThrough._events.data, undefined);
+      strictEqual(passThrough._readableState.pipes.length, 0);
+      passThrough.write('foobar');
+    }
+  }
+};
+
+export const pipeNeedDrain = {
+  async test(ctrl, env, ctx) {
+    // Pipe should pause temporarily if writable needs drain.
+    const w = new Writable({
+      write(buf, encoding, callback) {
+        queueMicrotask(callback);
+      },
+      highWaterMark: 1
+    });
+    while (w.write('asd'));
+    strictEqual(w.writableNeedDrain, true);
+    const r = new Readable({
+      read() {
+        this.push('asd');
+        this.push(null);
+      }
+    });
+    const pauseCalled = deferredPromise();
+    const endCalled = deferredPromise();
+    let pauseCount = 0;
+    r.on('pause', () => {
+      if (++pauseCount === 2) pauseCalled.resolve();
+    });
+    r.on('end', endCalled.resolve);
+    r.pipe(w);
+    await Promise.all([
+      pauseCalled.promise,
+      endCalled.promise,
+    ]);
+  }
+};
+
+export const pipeMultiplePipes = {
```

Comment:
Currently this test is doing the following
```
              ┌────────┐
    ┌────────►│Writable│
    │         └────────┘
    │
┌───┴─────┐   ┌────────┐
│Readable ├──►│Writable│
└───┬─────┘   └────────┘
    │
    │         ┌────────┐
    └────────►│Writable│
              └────────┘
```
Still going through this PR, but it so far does not have a test for long chains of piped streams.
```
┌─────────┐  ┌──────┐  ┌──────┐ ...  ┌──────┐
│Readable ├─►│Duplex├─►│Duplex├─────►│Duplex│
└─────────┘  └──────┘  └──────┘      └──────┘
```